### PR TITLE
Refactor done workspaces

### PR DESCRIPTION
**Pull request by ruuushhh**

### CODE REVIEW

File: apps/workspaces/actions.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+from apps.fyle.models import ExpenseGroupSettings
+from .utils import generate_qbo_refresh_token, assert_valid
+from .serializers import WorkspaceSerializer, QBOCredentialSerializer, \
+    WorkSpaceGeneralSettingsSerializer, LastExportDetailSerializer
+from .signals import post_delete_qbo_connection
+
+
+User = get_user_model()
+
+def qbo_workspace(user, org_id, org_name, org_currency):
```
Comment (ashwin1111) [inline apps/workspaces/actions.py:24]: update_or_create_workspace

File: apps/workspaces/views.py [hunk 2]
```diff
@@ -157,65 +136,19 @@ def get(self, request):
...
-
-            if workspace.onboarding_state == 'CONNECTION':
-                workspace.onboarding_state = 'MAP_EMPLOYEES'
-            workspace.save()
-
-            # Return the QBO credentials as serialized data
-            return Response(data=QBOCredentialSerializer(qbo_credentials).data, status=status.HTTP_200_OK)
-
+            return connect_qbo_oauth(refresh_token, realm_id, redirect_uri)
         except qbo_exc.UnauthorizedClientError:
```
Comment (ashwin1111) [inline apps/workspaces/views.py:152]: let's club all below exception and return a generic message

Comment (ashwin1111) [reply apps/workspaces/views.py:152]: make sure you do logger.info()

File: apps/workspaces/views.py [hunk 3]
```diff
@@ -328,59 +222,30 @@ def post(self, request, *args, **kwargs):
...
 
-class LastExportDetailView(viewsets.ViewSet):
+class LastExportDetailView(generics.RetrieveAPIView):
     """
     Last Export Details
     """
+    lookup_field = 'workspace_id'
+    lookup_field_kwarg = 'workspace_id'
+
+    quryset = LastExportDetail.objects.first()
```
Comment (ashwin1111) [inline apps/workspaces/views.py:232]: typo

Comment (ashwin1111) [reply apps/workspaces/views.py:232]: is it possible to add last_exported_at__isnull=False and total_expense_groups_count > 0?

File: apps/workspaces/views.py [hunk 4]
```diff
@@ -328,59 +222,30 @@ def post(self, request, *args, **kwargs):
...
-                'email': admin.email
-            })
-
         return Response(
                 data=admin_email,
                 status=status.HTTP_200_OK
             )
 
-class SetupE2ETestView(viewsets.ViewSet):
+class SetupE2ETestView(generics.CreateAPIView):
```
Comment (ashwin1111) [inline apps/workspaces/views.py:248]: can you move whole logic to actions?

File: apps/workspaces/views.py (hunk 3 above)
Comment (ruuushhh) [reply apps/workspaces/views.py]: not with RetrieveAPIView

Comment (ashwin1111) [reply apps/workspaces/views.py]: Screenshot 2023-06-26 at 5 04 45 PM
try this out @ruuushhh

File: apps/workspaces/views.py [hunk 5]
```diff
@@ -56,39 +42,14 @@ def post(self, request):
         """
         Create a Workspace
         """
-
         access_token = request.META.get('HTTP_AUTHORIZATION')
+        user = request.user
         fyle_user = get_fyle_admin(access_token.split(' ')[1], None)
```
Comment (ashwin1111) [inline apps/workspaces/views.py:47]: move all these logic to actions, there shouldn't be anything happening in views

Comment (ruuushhh) [reply apps/workspaces/views.py:47]: tried with this only @ashwin1111 but that didn't work
So had to get `request.META.get('HTTP_AUTHORIZATION')` from outside

File: apps/workspaces/views.py [hunk 6]
```diff
@@ -157,100 +102,24 @@ def get(self, request):
...
-            return Response({'message': 'QBO Application not found'}, status=status.HTTP_404_NOT_FOUND)
-
-        except qbo_exc.WrongParamsError as e:
-            return Response(json.loads(e.response), status=status.HTTP_400_BAD_REQUEST)
-
-        except qbo_exc.InternalServerError:
-            return Response({'message': 'Wrong/Expired Authorization code'}, status=status.HTTP_401_UNAUTHORIZED)
-
-
-    def patch(self, request, **kwargs):
```
Comment (ashwin1111) [inline apps/workspaces/views.py:232]: we'll need patch method to do post_delete_qbo_connection and revoke_refresh_token. Can do something similar to this, and then call the couple of methods
Screenshot 2023-07-06 at 12 30 04 PM

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**